File: tools/sit/run-fps-batch.js
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) SonarSource Sàrl
 * mailto:info AT sonarsource DOT com
 *
 * You can redistribute and/or modify this program under the terms of
 * the Sonar Source-Available License Version 1, as published by SonarSource Sàrl.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
import { spawnSync } from 'node:child_process';
import { mkdir, readFile, stat, writeFile } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import {
  isMain,
  parseOptionArgs,
  requireOption,
  resolvePathUnder,
  writeJsonFile,
} from './common.js';
import { summarizeFpsTsv } from './summarize-fps-results.js';

const FPS_NO_ISSUES_EXIT_CODE = 2;

export function parseRuleKeys(rulesText) {
  const seen = new Set();
  const ruleKeys = [];
  for (const rawLine of rulesText.split('\n')) {
    const ruleKey = rawLine.trim();
    if (!ruleKey || seen.has(ruleKey)) {
      continue;
    }
    seen.add(ruleKey);
    ruleKeys.push(ruleKey);
  }
  return ruleKeys;
}

export function sanitizeRuleKey(ruleKey) {
  return ruleKey.replaceAll(/[:/]/g, '__');
}

export function resolveRspecRefForRule(ruleKey, rspecRefs) {
  const repository = ruleKey.split(':', 1)[0];
  if (repository === 'javascript' || repository === 'typescript') {
    return rspecRefs.javascript;
  }
  if (repository === 'css') {
    return rspecRefs.css;
  }
  return undefined;
}

export async function runOneRule({
  fpsProjectDir,
  runDir,
  sitInput,
  ruleKey,
  outputPrefix,
  rspecRef,
  runCommand = defaultRunCommand,
}) {
  const command = [
    'uv',
    'run',
    '--project',
    fpsProjectDir,
    'fps',
    '--source',
    'sit',
    '--sit-input',
    sitInput,
    '--limit',
    '1000',
    '--export-json',
    '--output-prefix',
    outputPrefix,
    ruleKey,
  ];

  let completed;
  try {
    completed = runCommand(command, runDir, buildRunEnvironment(rspecRef));
  } catch (error) {
    console.error(`Failed to run FPS for ${ruleKey}: ${error.message ?? error}`);
    return 1;
  }

  if (completed.stdout) {
    process.stdout.write(completed.stdout);
  }
  if (completed.stderr) {
    process.stderr.write(completed.stderr);
  }

  const status = completed.status ?? 1;
  if (status === FPS_NO_ISSUES_EXIT_CODE) {
    await writeEmptyReport(join(runDir, 'reports', `${outputPrefix}_summary.json`));
    return 0;
  }
  return status;
}

export async function writeEmptyReport(reportPath) {
  await writeJsonFile(reportPath, {
    metrics: {
      total_issues_analyzed: 0,
      false_positive_rate: 0.0,
      total_clusters: 0,
    },
    clusters: [],
  });
}

function defaultRunCommand(command, cwd, env) {
  return spawnSync(command[0], command.slice(1), {
    cwd,
    env,
    encoding: 'utf8',
    stdio: ['ignore', 'pipe', 'pipe'],
  });
}

function buildRunEnvironment(rspecRef) {
  if (!rspecRef) {
    return process.env;
  }

  return {
    ...process.env,
    RSPEC_GITHUB_BRANCH: rspecRef,
  };
}

function utcNow() {
  return new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
}

async function main() {
  const args = parseOptionArgs(process.argv);
  const workspace = requireOption(args, '--workspace');
  const safeWorkspace = resolvePathUnder('/', workspace, '--workspace');
  const fpsProjectDir = join(safeWorkspace, 'dev-tools', 'scripts');
  const sitInput = resolvePathUnder(
    safeWorkspace,
    requireOption(args, '--sit-input'),
    '--sit-input',
  );
  const reportsDir = resolvePathUnder(
    safeWorkspace,
    requireOption(args, '--reports-dir'),
    '--reports-dir',
  );
  const summaryTsv = resolvePathUnder(
    safeWorkspace,
    requireOption(args, '--summary-tsv'),
    '--summary-tsv',
  );
  const summaryJson = resolvePathUnder(
    safeWorkspace,
    requireOption(args, '--summary-json'),
    '--summary-json',
  );
  const runDir = resolvePathUnder(safeWorkspace, dirname(reportsDir), 'run directory');
  const rspecRefs = await loadRspecRefs(safeWorkspace);

  await requireDirectory(fpsProjectDir, 'FPS project directory');
  await requireDirectory(sitInput, 'SIT input directory');

  const ruleKeys = parseRuleKeys(requireOption(args, '--rules-text'));
  await mkdir(runDir, { recursive: true });
  await mkdir(dirname(summaryTsv), { recursive: true });

  const rows = [];
  let hasFailures = false;
  for (const ruleKey of ruleKeys) {
    const outputPrefix = `pr${requireOption(args, '--pr-number')}_${requireOption(
      args,
      '--head-sha',
    ).slice(0, 12)}_${sanitizeRuleKey(ruleKey)}`;
    const startedAt = utcNow();
    const exitCode = await runOneRule({
      fpsProjectDir,
      runDir,
      sitInput,
      ruleKey,
      outputPrefix,
      rspecRef: resolveRspecRefForRule(ruleKey, rspecRefs),
    });
    const finishedAt = utcNow();
    const status = exitCode === 0 ? 'success' : 'failure';
    if (exitCode !== 0) {
      hasFailures = true;
    }
    rows.push([ruleKey, outputPrefix, status, String(exitCode), startedAt, finishedAt].join('\t'));
  }

  await writeFile(summaryTsv, rows.length === 0 ? '' : `${rows.join('\n')}\n`, 'utf8');
  const payload = await summarizeFpsTsv({
    inputTsv: summaryTsv,
    outputJson: summaryJson,
    reportsDir,
    allowedRoot: safeWorkspace,
  });
  console.log(
    `Wrote FPS summary to ${summaryJson} (total=${payload.total}, failed=${payload.failed})`,
  );
  if (hasFailures) {
    throw new Error('One or more FPS runs failed');
  }
}

async function requireDirectory(path, label) {
  try {
    if ((await stat(path)).isDirectory()) {
      return;
    }
  } catch (error) {
    if (error.code !== 'ENOENT') {
      throw error;
    }
  }
  throw new Error(`${label} does not exist: ${path}`);
}

async function loadRspecRefs(workspace) {
  return {
    javascript: await readOptionalTrimmedFile(
      join(workspace, 'sonar-plugin', 'javascript-checks', 'src', 'main', 'resources', 'rspec.sha'),
    ),
    css: await readOptionalTrimmedFile(
      join(workspace, 'sonar-plugin', 'css', 'src', 'main', 'resources', 'rspec.sha'),
    ),
  };
}

async function readOptionalTrimmedFile(path) {
  try {
    const value = (await readFile(path, 'utf8')).trim();
    return value === '' ? undefined : value;
  } catch (error) {
    if (error.code === 'ENOENT') {
      return undefined;
    }
    throw error;
  }
}

if (isMain(import.meta.url)) {
  main().catch(error => {
    console.error(error instanceof Error ? error.message : error);
    process.exit(1);
  });
}

File: dev-tools/scripts/sit_pr_fps/run_fps_batch.py
"""Run FPS sequentially for all changed rules and persist run summaries.

Executes `uv run --project dev-tools/scripts fps` per rule key, records
per-rule status in a TSV file, then materializes a consolidated JSON summary
for publish steps.
"""

import argparse
import json
import re
import subprocess
import sys
from datetime import UTC, datetime
from pathlib import Path

from common import resolve_path_under
from summarize_fps_results import summarize_tsv

FPS_NO_ISSUES_EXIT_CODE = 2


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Run FPS for a batch of changed rules")
    parser.add_argument("--rules-text", required=True, help="Newline-separated full rule keys")
    parser.add_argument("--workspace", required=True, help="GitHub workspace root")
    parser.add_argument("--sit-input", required=True, help="Normalized SIT bundle input directory")
    parser.add_argument("--reports-dir", required=True, help="FPS reports directory")
    parser.add_argument("--summary-tsv", required=True, help="Per-rule run summary TSV path")
    parser.add_argument("--summary-json", required=True, help="Per-rule run summary JSON path")
    parser.add_argument("--pr-number", required=True)
    parser.add_argument("--head-sha", required=True)
    return parser.parse_args()


def utc_now() -> str:
    return datetime.now(UTC).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def parse_rule_keys(rules_text: str) -> list[str]:
    seen: set[str] = set()
    rule_keys: list[str] = []
    for raw_line in rules_text.splitlines():
        rule_key = raw_line.strip()
        if not rule_key or rule_key in seen:
            continue
        seen.add(rule_key)
        rule_keys.append(rule_key)
    return rule_keys


def sanitize_rule_key(rule_key: str) -> str:
    return re.sub(r"[:/]", "__", rule_key)


def run_one_rule(fps_project_dir: Path, run_dir: Path, sit_input: Path, rule_key: str, output_prefix: str) -> int:
    command = [
        "uv",
        "run",
        "--project",
        str(fps_project_dir),
        "fps",
        "--source",
        "sit",
        "--sit-input",
        str(sit_input),
        "--limit",
        "1000",
        "--export-json",
        "--output-prefix",
        output_prefix,
        rule_key,
    ]
    try:
        completed = subprocess.run(command, cwd=run_dir, check=False, capture_output=True, text=True)
    except OSError as err:
        print(f"Failed to run FPS for {rule_key}: {err}", file=sys.stderr)
        return 1

    if completed.stdout:
        print(completed.stdout, end="")
    if completed.stderr:
        print(completed.stderr, end="", file=sys.stderr)

    if completed.returncode == FPS_NO_ISSUES_EXIT_CODE:
        write_empty_report(run_dir / "reports" / f"{output_prefix}_summary.json")
        return 0

    return completed.returncode


def write_empty_report(report_path: Path) -> None:
    report_path.parent.mkdir(parents=True, exist_ok=True)
    report_path.write_text(
        json.dumps(
            {
                "metrics": {
                    "total_issues_analyzed": 0,
                    "false_positive_rate": 0.0,
                    "total_clusters": 0,
                },
                "clusters": [],
            },
            indent=2,
        ),
        encoding="utf-8",
    )


def resolve_run_dir(workspace: Path, reports_dir: Path) -> Path:
    return resolve_path_under(workspace, reports_dir.parent, "run directory")


def main() -> int:
    args = parse_args()
    workspace = Path(args.workspace).resolve()
    fps_project_dir = workspace / "dev-tools" / "scripts"
    sit_input = resolve_path_under(workspace, args.sit_input, "--sit-input")
    reports_dir = resolve_path_under(workspace, args.reports_dir, "--reports-dir")
    run_dir = resolve_run_dir(workspace, reports_dir)
    summary_tsv = resolve_path_under(workspace, args.summary_tsv, "--summary-tsv")
    summary_json = resolve_path_under(workspace, args.summary_json, "--summary-json")

    if not fps_project_dir.is_dir():
        print(f"FPS project directory does not exist: {fps_project_dir}", file=sys.stderr)
        return 1
    if not sit_input.is_dir():
        print(f"SIT input directory does not exist: {sit_input}", file=sys.stderr)
        return 1

    rule_keys = parse_rule_keys(args.rules_text)
    run_dir.mkdir(parents=True, exist_ok=True)
    summary_tsv.parent.mkdir(parents=True, exist_ok=True)

    failures = 0
    with open(summary_tsv, "w", encoding="utf-8") as summary_file:
        for rule_key in rule_keys:
            safe_rule_key = sanitize_rule_key(rule_key)
            output_prefix = f"pr{args.pr_number}_{args.head_sha[:12]}_{safe_rule_key}"
            started_at = utc_now()
            exit_code = run_one_rule(
                fps_project_dir=fps_project_dir,
                run_dir=run_dir,
                sit_input=sit_input,
                rule_key=rule_key,
                output_prefix=output_prefix,
            )
            finished_at = utc_now()
            status = "success" if exit_code == 0 else "failure"
            if exit_code != 0:
                failures = 1

            summary_file.write(f"{rule_key}\t{output_prefix}\t{status}\t{exit_code}\t{started_at}\t{finished_at}\n")

    payload = summarize_tsv(
        input_path=summary_tsv, output_path=summary_json, reports_dir=reports_dir, allowed_root=workspace
    )
    print(f"Wrote FPS summary to {summary_json} (total={payload['total']}, failed={payload['failed']})")

    if failures != 0:
        print("One or more FPS runs failed", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
